File: results/dnd-rest-benchmark/lifecycle-runs/20260730T075409Z_pi_kimi-k2p7-code_ruby-stdlib/lib/sessions.rb
# frozen_string_literal: true

require 'json'
require_relative 'persistence'

# Session scheduling: campaign sessions and attendance.
module Sessions
  ISO8601_RE = /\A\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z\z/.freeze

  def self.create(campaign_id, payload)
    return [:invalid] unless campaign_id.is_a?(String) && !campaign_id.empty?

    data = validate_session(payload)
    return [:invalid] unless data

    Persistence.db do |d|
      campaign_exists = d.get_first_value('SELECT 1 FROM campaigns WHERE id = ?', campaign_id)
      next [:not_found] unless campaign_exists

      existing = d.get_first_value('SELECT 1 FROM campaign_sessions WHERE id = ?', data[:id])
      next [:conflict] if existing

      d.execute(
        'INSERT INTO campaign_sessions (id, campaign_id, starts_at, duration_minutes, agenda_json) VALUES (?, ?, ?, ?, ?)',
        [data[:id], campaign_id, data[:starts_at], data[:duration_minutes], JSON.generate(data[:agenda])]
      )

      [:ok, {
        'id' => data[:id],
        'starts_at' => data[:starts_at],
        'duration_minutes' => data[:duration_minutes],
        'agenda_count' => data[:agenda].length
      }]
    end
  end

  def self.record_attendance(campaign_id, session_id, payload)
    return [:invalid] unless campaign_id.is_a?(String) && !campaign_id.empty? &&
                             session_id.is_a?(String) && !session_id.empty?

    present = payload['present']
    absent = payload['absent']
    return [:invalid] unless present.is_a?(Array) && absent.is_a?(Array) &&
                             present.all? { |c| c.is_a?(String) && !c.empty? } &&
                             absent.all? { |c| c.is_a?(String) && !c.empty? }

    Persistence.db do |d|
      campaign_exists = d.get_first_value('SELECT 1 FROM campaigns WHERE id = ?', campaign_id)
      next [:not_found] unless campaign_exists

      session_exists = d.get_first_value(
        'SELECT 1 FROM campaign_sessions WHERE id = ? AND campaign_id = ?',
        [session_id, campaign_id]
      )
      next [:not_found] unless session_exists

      d.execute('DELETE FROM session_attendance WHERE session_id = ?', session_id)

      records = {}
      present.each { |c| records[c] = 'present' }
      absent.each { |c| records[c] = 'absent' }
      records.each do |character_id, status|
        d.execute(
          'INSERT INTO session_attendance (session_id, character_id, status) VALUES (?, ?, ?)',
          [session_id, character_id, status]
        )
      end

      [:ok, {
        'session_id' => session_id,
        'present_count' => present.length,
        'absent_count' => absent.length
      }]
    end
  end

  # Note: returns :not_found (not :invalid) for malformed IDs. This preserves
  # the behavior exercised by the cumulative session-scheduling suite.
  def self.next_session(campaign_id)
    return [:not_found] unless campaign_id.is_a?(String) && !campaign_id.empty?

    Persistence.db do |d|
      campaign_exists = d.get_first_value('SELECT 1 FROM campaigns WHERE id = ?', campaign_id)
      next [:not_found] unless campaign_exists

      row = d.get_first_row(
        'SELECT id, starts_at, agenda_json FROM campaign_sessions WHERE campaign_id = ? ORDER BY starts_at ASC LIMIT 1',
        campaign_id
      )
      next [:not_found] unless row

      id, starts_at, agenda_json = row
      agenda = JSON.parse(agenda_json) rescue []

      [:ok, {
        'id' => id,
        'starts_at' => starts_at,
        'agenda_count' => agenda.length
      }]
    end
  end

  def self.validate_session(payload)
    return nil unless payload.is_a?(Hash)

    id = payload['id']
    starts_at = payload['starts_at']
    duration_minutes = payload['duration_minutes']
    agenda = payload['agenda']

    return nil unless id.is_a?(String) && !id.empty? &&
                      starts_at.is_a?(String) && !starts_at.empty? && starts_at.match?(ISO8601_RE) &&
                      duration_minutes.is_a?(Integer) && duration_minutes > 0 &&
                      agenda.is_a?(Array) &&
                      agenda.all? { |a| a.is_a?(String) && !a.empty? }

    {
      id: id,
      starts_at: starts_at,
      duration_minutes: duration_minutes,
      agenda: agenda
    }
  end
  private_class_method :validate_session
end

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T072747Z_pi_kimi-k2p7-code_typescript-vite/src/lib/db/sessions.ts
// Game session scheduling and attendance persistence.

import { db } from './connection.js';
import type { GameSession } from '../types.js';

export function createSession(session: GameSession): void {
  db.prepare('INSERT INTO sessions (id, campaign_id, starts_at, duration_minutes, agenda) VALUES (?, ?, ?, ?, ?)').run(
    session.id,
    session.campaign_id,
    session.starts_at,
    session.duration_minutes,
    JSON.stringify(session.agenda),
  );
}

export function sessionExists(id: string): boolean {
  const row = db.prepare('SELECT 1 FROM sessions WHERE id = ?').get(id) as { '1': number } | undefined;
  return row !== undefined;
}

export function getSession(id: string): GameSession | undefined {
  const row = db.prepare('SELECT id, campaign_id, starts_at, duration_minutes, agenda FROM sessions WHERE id = ?').get(id) as
    | { id: string; campaign_id: string; starts_at: string; duration_minutes: number; agenda: string }
    | undefined;
  if (!row) return undefined;
  return {
    id: row.id,
    campaign_id: row.campaign_id,
    starts_at: row.starts_at,
    duration_minutes: row.duration_minutes,
    agenda: JSON.parse(row.agenda) as string[],
  };
}

export function getSessionsByCampaign(campaignId: string): GameSession[] {
  const rows = db
    .prepare('SELECT id, campaign_id, starts_at, duration_minutes, agenda FROM sessions WHERE campaign_id = ? ORDER BY starts_at, id')
    .all(campaignId) as {
    id: string;
    campaign_id: string;
    starts_at: string;
    duration_minutes: number;
    agenda: string;
  }[];
  return rows.map((row) => ({
    id: row.id,
    campaign_id: row.campaign_id,
    starts_at: row.starts_at,
    duration_minutes: row.duration_minutes,
    agenda: JSON.parse(row.agenda) as string[],
  }));
}

export function recordAttendance(sessionId: string, present: string[], absent: string[]): void {
  const insert = db.prepare('INSERT OR REPLACE INTO attendance (session_id, character_id, present) VALUES (?, ?, ?)');
  for (const characterId of present) {
    insert.run(sessionId, characterId, 1);
  }
  for (const characterId of absent) {
    insert.run(sessionId, characterId, 0);
  }
}

export function getAttendanceBySession(sessionId: string): { present: string[]; absent: string[] } {
  const rows = db.prepare('SELECT character_id, present FROM attendance WHERE session_id = ?').all(sessionId) as {
    character_id: string;
    present: number;
  }[];
  const present: string[] = [];
  const absent: string[] = [];
  for (const row of rows) {
    if (row.present) {
      present.push(row.character_id);
    } else {
      absent.push(row.character_id);
    }
  }
  return { present, absent };
}
